### PR TITLE
feat: connect team management page to backend

### PR DESCRIPTION
**Pull request by iraboldyrieva**
## Describe your changes

- Integration of the Team Management page with backend.
- Implemented functionalities:
  - Pagination.
  - Sorting by role.
  - Filtering by role.
  - Adding users.
  - Updating users.
  - Deleting users.

## Issue ticket code (and/or) and link

- [SCRUM-28: As admin I want to be able to manage users](https://codecraftersintership.atlassian.net/browse/SCRUM-28)

### **General**

- [x] Assigned myself to the PR
- [x] Assigned the appropriate labels to the PR
- [x] Assigned the appropriate reviewers to the PR
- [x] Updated the documentation
- [x] Performed a self-review of my code
- [x] Types for input and output parameters
- [x] Don't have "any" on my code
- [x] Used the try/catch pattern for error handling
- [x] Don't have magic numbers
- [x] Compare only with constants not with strings
- [x] No ternary operator inside the ternary operator
- [x] Don't have commented code
- [x] no links in the code, env links should be in env file (for example: server url), constant links (for example default avatar URL) should be in constant file.
- [x] Used camelCase for variables and functions
- [x] Date and time formats are on the constants
- [x] Functions are public only if it's used outside the class
- [x] No hardcoded values
- [x] covered by tests
- [x] Check your commit messages meet the [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/).

### Frontend

- [x] Components and business logic are separated
- [x] Colors, Font Size, and Font Name is on the theme or in the constants
- [x] No text in the components, use i18n approach
- [x] No inline styles
- [x] Imports are absolute
- [x] Attach a screenshot if PR has visual changes.

### CODE REVIEW

File: src/components/UserItem/index.tsx [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+const UserItem: React.FC<UserItemProps> = ({
+  user,
+  fetchUsers,
+  addUserToList,
+}) => {
+  const [isPopupOpen, setIsPopupOpen] = useState(false);
+  const firstName = user.fullName.split(' ')[0] || '';
+  const lastName = user.fullName.split(' ')[1] || '';
```
Comment (lyudad) [inline src/components/UserItem/index.tsx]: create function in utility - getFirstName and getSeconName - assume it could be more then one place where this functionality could be needed

File: src/components/UserItem/index.tsx
Comment (lyudad) [inline src/components/UserItem/index.tsx:1]: can we this about better name?

File: src/pages/TeamManagment/UserForm/UserForm.tsx [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+      }
+
+      closeModal();
+    } catch (error) {
+      console.log(error);
+      toast(t('settings.message.error'), { type: 'error' });
+    }
+  };
+
+  return (
```
Comment (lyudad) [inline src/pages/TeamManagment/UserForm/UserForm.tsx:27]: all above move to hook

File: src/pages/TeamManagment/UserForm/styles.css
Comment (lyudad) [inline src/pages/TeamManagment/UserForm/styles.css:1]: can we avoid messed up styled component with css?

File: src/pages/TeamManagment/index.tsx [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+          sx={{ color: COLORS.text.dark, fontSize: FONT.fontSize.small }}
+        >
+          {t('settings.columns.actions')}
+        </Typography>
+      </Box>
+
+      <Divider />
+
+      {loading ? (
+        <Loader />
```
Comment (lyudad) [inline src/pages/TeamManagment/index.tsx]: better practice make loader as wrapper. so you make wrapper and props = isLoading, and inside wrapper component you sert condition or show loader or show {child}

File: src/pages/TeamManagment/useFetchUsers.tsx [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        },
+      });
+
+      const adaptedUsers = response.data.users.map((user: ApiUser) => {
+        const [firstName, ...lastNameParts] = user.full_name.split(' ');
+        return {
+          ...user,
+          firstName,
+          lastName: lastNameParts.join(' ') || '',
+          phoneNumber: user.phone_number || 'No phone number',
```
Comment (lyudad) [inline src/pages/TeamManagment/useFetchUsers.tsx]: 1.3.3

File: src/components/DriverAvatar/index.tsx
Comment (lyudad) [inline src/components/DriverAvatar/index.tsx:1]: why it called Driver Avatar? I think we have similar approach in full app

**Comment by iraboldyrieva** [comment]
and i forget to add screens
Знімок екрана 2024-11-25 о 17 20 32
Знімок екрана 2024-11-25 о 17 20 44
Знімок екрана 2024-11-25 о 17 20 55
Знімок екрана 2024-11-25 о 17 21 08
Знімок екрана 2024-11-25 о 17 21 20

File: src/pages/TeamManagment/UserForm/useUserForm.tsx [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+      return decodedToken;
+    } catch (error) {
+      console.error('Error decoding token:', error);
+      return null;
+    }
+  };
+
+  const sendData = async (formData: UserFormInputs) => {
+    try {
+      const decodedToken = authToken ? decodeToken(authToken) : null;
```
Comment (lyudad) [inline src/pages/TeamManagment/UserForm/useUserForm.tsx]: why you need it? it look like some overwork

File: src/pages/TeamManagment/UserForm/useUserForm.tsx [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+      if (!decodedToken) {
+        console.error('Invalid token');
+        toast(t('settings.message.invalidToken'), { type: 'error' });
+        return;
+      }
+
+      const companyId = decodedToken.company_id.id;
```
Comment (lyudad) [inline src/pages/TeamManagment/UserForm/useUserForm.tsx]: why you take company id from token? if user belong to company and you need access to this value from some places, save this info during login in redux

File: src/pages/TeamManagment/UserForm/useUserForm.tsx (hunk 5 above)
Comment (iraboldyrieva) [reply src/pages/TeamManagment/UserForm/useUserForm.tsx]: fixed it

File: src/pages/TeamManagment/UserForm/useUserForm.tsx (hunk 6 above)
Comment (iraboldyrieva) [reply src/pages/TeamManagment/UserForm/useUserForm.tsx]: fixed it

File: src/api/userActions.ts [hunk 7]
```diff
@@ -39,7 +47,13 @@ export const verifyToken =
...
 
       if (token) {
-        dispatch(setAccessToken({ token, role }));
+        const decodedToken = jwtDecode<DecodedToken>(token);
+
+        console.log('Decoded Token:', decodedToken);
+
+        dispatch(
+          setAccessToken({ token, role, companyId: decodedToken.company_id.id })
+        );
```
Comment (lyudad) [inline src/api/userActions.ts]: и что поменялось? с токена информацию НЕ берем! есть случаи когда это нужно, это не тот случай

Comment (iraboldyrieva) [reply src/api/userActions.ts]: This token it's not my code, it's approved code of Oleg. I'm just added there company id.

File: src/api/userActions.ts [hunk 8]
```diff
@@ -1,10 +1,16 @@
-import { toast } from 'react-toastify';
-import { Dispatch } from '@reduxjs/toolkit';
 import axios from 'axios';
-import i18n from '@/utils/i18n';
+import { toast } from 'react-toastify';
 
-import axiosInstance from '@/utils/axiosInstance';
 import { setAccessToken } from '@/store/slices/authSlice';
```
Comment (Parle648) [inline src/api/userActions.ts:5]: 1.1.14

File: src/components/ModalForm/index.tsx [hunk 9]
```diff
@@ -1,12 +1,12 @@
...
 
 import Button from '@/components/Button';
 import {
-  formWrapper,
-  formContainer,
   EditIcon,
+  formContainer,
+  formWrapper,
 } from '@/components/ModalForm/styles';
+import { Box, Typography } from '@mui/material';
```
Comment (Parle648) [inline src/components/ModalForm/index.tsx:9]: 1.1.14

File: src/components/UserAvatar/index.tsx [hunk 10]
```diff
@@ -9,7 +9,10 @@ interface DriverAvatarProps {
   lastName?: string;
 }
 
-const DriverAvatar: React.FC<DriverAvatarProps> = ({ firstName, lastName }) => {
+const UserAvatar: React.FC<DriverAvatarProps> = ({
```
Comment (Parle648) [inline src/components/UserAvatar/index.tsx]: we are already using the name DriverAvatar. everything will break.

File: src/components/UserRow/index.tsx [hunk 11]
```diff
@@ -0,0 +1,129 @@
+import { t } from 'i18next';
+import React, { useState } from 'react';
+import { toast } from 'react-toastify';
+
+import { assets } from '@/assets/assets';
+import DriverAvatar from '@/components/UserAvatar';
```
Comment (Parle648) [inline src/components/UserRow/index.tsx]: you replaced DriverAvatar by UserAvatar and don't replace it here (it should be in a separate pr with refactoring not here)

File: src/components/UserRow/index.tsx [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+import { COLORS } from '@/constants/colors';
+import { User } from '@/pages/TeamManagment/types';
+import UserForm from '@/pages/TeamManagment/UserForm/UserForm';
+import { deleteUser } from '@/services/usersService';
+import { getFirstName, getSecondName } from '@/utils/nameUtils';
+import { Box, Divider, IconButton, Typography } from '@mui/material';
```
Comment (Parle648) [inline src/components/UserRow/index.tsx]: 1.1.14

File: src/components/UserRow/index.tsx [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+  fetchUsers: () => void;
+  addUserToList: (user: User) => void;
+}
+
+const UserRow: React.FC<UserItemProps> = ({
+  user,
+  fetchUsers,
+  addUserToList,
+}) => {
+  const [isPopupOpen, setIsPopupOpen] = useState(false);
```
Comment (Parle648) [inline src/components/UserRow/index.tsx:33]: I'd better make like this useState<boolean>(false)

File: src/components/UserRow/index.tsx [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+  const firstName = getFirstName(user.fullName);
+  const lastName = getSecondName(user.fullName);
+
+  const handleDelete = async () => {
+    try {
+      await deleteUser(user.id);
+      fetchUsers();
+    } catch (error) {
+      toast(t('settings.message.errorDelete'), { type: 'error' });
+      console.error('Failed to delete user:', error);
```
Comment (Parle648) [inline src/components/UserRow/index.tsx]: 1.1.8

File: src/components/UserRow/index.tsx [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+const UserRow: React.FC<UserItemProps> = ({
+  user,
+  fetchUsers,
+  addUserToList,
+}) => {
+  const [isPopupOpen, setIsPopupOpen] = useState(false);
+  const firstName = getFirstName(user.fullName);
+  const lastName = getSecondName(user.fullName);
+
+  const handleDelete = async () => {
```
Comment (Parle648) [inline src/components/UserRow/index.tsx]: 1.1.1

File: src/components/UserRow/index.tsx [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+      toast(t('settings.message.errorDelete'), { type: 'error' });
+      console.error('Failed to delete user:', error);
+    }
+    setIsPopupOpen(false);
+  };
+
+  return (
+    <>
+      <Box
+        sx={{
```
Comment (Parle648) [inline src/components/UserRow/index.tsx:51]: 1.3.4

File: src/pages/CreateRoute/DriversStage/components/DriverCard/index.tsx [hunk 17]
```diff
@@ -1,12 +1,12 @@
-import { FC } from 'react';
-import { Box, Typography } from '@mui/material';
 import { t } from 'i18next';
+import { FC } from 'react';
```
Comment (Parle648) [inline src/pages/CreateRoute/DriversStage/components/DriverCard/index.tsx]: react on the first line

File: src/pages/CreateRoute/DriversStage/components/DriverCard/index.tsx [hunk 18]
```diff
@@ -1,12 +1,12 @@
...
 
-import DriverAvatar from '@/components/DriverAvatar/index.tsx';
-import { Customer } from '@/interfaces/interfaces.ts';
 import CustomCheckbox from '@/components/Checkbox/index.tsx';
+import DriverAvatar from '@/components/UserAvatar/index.tsx';
+import { Customer } from '@/interfaces/interfaces.ts';
+import { Box, Typography } from '@mui/material';
```
Comment (Parle648) [inline src/pages/CreateRoute/DriversStage/components/DriverCard/index.tsx]: 1.1.14

File: src/pages/Orders/components/OrderItem/OrderItem.tsx [hunk 19]
```diff
@@ -1,23 +1,23 @@
...
+  customerBlock,
+  luggageStyles,
+  orderRow,
   popup,
+  routeBlock,
+  statusBlock,
 } from '@/pages/Orders/components/OrderItem/styles';
-import { OrderStatuses } from '@/interfaces/interfaces';
+import { OrderItemInterface } from '@/pages/Orders/components/OrderItem/types';
+import { Box, Typography } from '@mui/material';
```
Comment (Parle648) [inline src/pages/Orders/components/OrderItem/OrderItem.tsx]: 1.1.14

File: src/pages/TeamManagment/UserForm/UserForm.tsx [hunk 20]
```diff
@@ -0,0 +1,128 @@
+import { t } from 'i18next';
+import React from 'react';
```
Comment (Parle648) [inline src/pages/TeamManagment/UserForm/UserForm.tsx:1]: first line

File: src/pages/TeamManagment/UserForm/UserForm.tsx [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+
+import { assets } from '@/assets/assets';
+import Button from '@/components/Button';
+import ModalForm from '@/components/ModalForm';
+import TextInput from '@/components/TextInput';
+import { MenuItem, Select, Typography } from '@mui/material';
```
Comment (Parle648) [inline src/pages/TeamManagment/UserForm/UserForm.tsx]: 1.1.14

File: src/pages/TeamManagment/UserForm/styles.ts [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+  },
+  '& .MuiInputLabel-root': {
+    top: '-8px',
+  },
+};
+
+export const wrapper: SxProps<Theme> = {
+  width: '100vw',
+  height: '100vh',
+  backgroundColor: '#ccccccc2',
```
Comment (Parle648) [inline src/pages/TeamManagment/UserForm/styles.ts]: 1.3.2

File: src/pages/TeamManagment/UserForm/useUserForm.tsx [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+import { useSelector } from 'react-redux';
+import { toast } from 'react-toastify';
+
+import { RootState } from '@/store/store';
+import axiosInstance from '@/utils/axiosInstance';
+import { getFirstName, getSecondName } from '@/utils/nameUtils';
+
+import { UserFormInputs, UserFormProps } from '../types';
+
+export const useUserForm = ({
```
Comment (Parle648) [inline src/pages/TeamManagment/UserForm/useUserForm.tsx:33]: 1.1.1

File: src/pages/TeamManagment/index.tsx [hunk 24]
```diff
@@ -0,0 +1,207 @@
+import { t } from 'i18next';
+import React from 'react';
```
Comment (Parle648) [inline src/pages/TeamManagment/index.tsx:1]: react - first line + 1.1.1

File: src/pages/TeamManagment/index.tsx [hunk 25]
```diff
@@ -0,0 +1,207 @@
...
+    searchTerm,
+    filterByRole,
+    sortOrder
+  );
+
+  return (
+    <>
+      <Loader />
+      <Box
+        sx={{
```
Comment (Parle648) [inline src/pages/TeamManagment/index.tsx:62]: 1.3.4

File: src/pages/TeamManagment/useFetchUsers.tsx [hunk 26]
```diff
@@ -0,0 +1,77 @@
+import { t } from 'i18next';
+import { useEffect, useState } from 'react';
```
Comment (Parle648) [inline src/pages/TeamManagment/useFetchUsers.tsx:1]: first line

File: src/pages/TeamManagment/useFetchUsers.tsx [hunk 27]
```diff
@@ -0,0 +1,77 @@
...
+import { useSelector } from 'react-redux';
+import { toast } from 'react-toastify';
+
+import { RootState } from '@/store/store';
+import axiosInstance from '@/utils/axiosInstance';
+
+import { ApiUser, User } from './types';
+
+const useFetchUsers = (
```
Comment (Parle648) [inline src/pages/TeamManagment/useFetchUsers.tsx:22]: 1.1.1

File: src/pages/TeamManagment/useFetchUsers.tsx [hunk 28]
```diff
@@ -0,0 +1,77 @@
...
+          lastName: lastNameParts.join(' ') || '',
+          phoneNumber: user.phone_number || t('setting.noPhone'),
+        };
+      });
+
+      setUsers(adaptedUsers);
+      setTotalPages(response.data.pagesCount);
+    } catch (err) {
+      toast.error(t('settings.message.fetchError'));
+      console.error(err);
```
Comment (Parle648) [inline src/pages/TeamManagment/useFetchUsers.tsx]: 1.1.8

File: src/pages/TeamManagment/useUserFilters.tsx [hunk 29]
```diff
@@ -0,0 +1,55 @@
+import React, { useState } from 'react';
+
+const useUserFilters = () => {
```
Comment (Parle648) [inline src/pages/TeamManagment/useUserFilters.tsx]: 1.1.1

File: src/pages/components/AdminListTable/index.tsx [hunk 30]
```diff
@@ -1,28 +1,31 @@
...
+  TableTitle,
 } from '@/pages/components/AdminListTable/styles';
 import DriverAvatar from '@/components/DriverAvatar';
 import { AdminListTableProps } from '@/interfaces/AdminList';
 import DeleteAdmin from '@/pages/components/DeleteAdmin';
-import AdminForm from '@/pages/components/AdminForm';
-import { addAdmin, updateAdmin } from '@/api/adminActions';
-import editIcon from '@/assets/edit.png';
-import { ITEMS_PER_PAGE_OPTIONS } from '@/constants/constants';
+import { Box, MenuItem, Pagination, Typography } from '@mui/material';
```
Comment (Parle648) [inline src/pages/components/AdminListTable/index.tsx]: 1.1.14

File: src/services/usersService.ts [hunk 31]
```diff
@@ -0,0 +1,27 @@
+import axiosInstance from '@/utils/axiosInstance';
+
+export const getUsers = async (
```
Comment (Parle648) [inline src/services/usersService.ts:22]: 1.1.1

File: src/store/slices/authSlice.ts [hunk 32]
```diff
@@ -1,17 +1,19 @@
-import { createSlice, PayloadAction } from '@reduxjs/toolkit';
 import { persistReducer } from 'redux-persist';
 
 import persistConfig from '@/store/persistConfig';
+import { createSlice, PayloadAction } from '@reduxjs/toolkit';
```
Comment (Parle648) [inline src/store/slices/authSlice.ts]: 1.1.14

**Review by Parle648 — COMMENTED** [review]
Left comments

File: src/components/UserAvatar/index.tsx (hunk 10 above)
Comment (iraboldyrieva) [reply src/components/UserAvatar/index.tsx]: Lyuda told me to change it for more reusable name

File: src/pages/TeamManagment/UserForm/useUserForm.tsx [hunk 33]
```diff
@@ -0,0 +1,141 @@
...
+        };
+
+        addUserToList(adaptedUser);
+      } else {
+        fetchUsers();
+      }
+
+      closeModal();
+    } catch (error) {
+      console.error(error);
```
Comment (oyakovytskyi) [inline src/pages/TeamManagment/UserForm/useUserForm.tsx:124]: 1.1.8

File: src/pages/TeamManagment/index.tsx [hunk 34]
```diff
@@ -0,0 +1,207 @@
...
```
Comment (AlexeyBezrebriy) [inline src/pages/TeamManagment/index.tsx:62]: 1.3.4

**Review by oyakovytskyi — APPROVED** [review]
